src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala

```diff
@@ -0,0 +1,223 @@
+  EndOfFileException, UserInterruptException, EOFError,
+  History, LineReader, ParsedLine, Parser}, Parser.ParseContext
+import org.jline.reader.impl.DefaultParser
+
+import java.util.{List => JList}
+
+/** A Reader that delegates to JLine3.
+ */
+class Reader private (config: ShellConfig, reader: LineReader, val accumulator: Accumulator, val completion: shell.Completion) extends shell.InteractiveReader {
+  val history: shell.History = new HistoryAdaptor(reader.getHistory)
+  def interactive: Boolean = true
+  protected def readOneKey(prompt: String): Int = ???
+  protected def readOneLine(prompt: String): String = {
+    try {
+      reader.readLine(prompt)
+    } catch {
+      case _: EndOfFileException | _: UserInterruptException => reader.getBuffer.delete() ; null
+    }
+  }
+  def redrawLine(): Unit = ???
+  def reset(): Unit = accumulator.reset()
+}
+
+object Reader {
+  import org.jline.reader.LineReaderBuilder
+  import org.jline.reader.impl.history.DefaultHistory
+  import org.jline.terminal.TerminalBuilder
+
+  /** Construct a Reader with various JLine3-specific set-up.
+   *  The `shell.Completion` is wrapped in the `jline.Completion` bridge to enable completion from JLine3.
+   */
+  def apply(config: ShellConfig, repl: Repl, completion: shell.Completion, accumulator: Accumulator): Reader = {
+    require(repl != null)
+    if (config.isReplDebug) initLogging()
+
+    System.setProperty(LineReader.PROP_SUPPORT_PARSEDLINE, java.lang.Boolean.TRUE.toString())
+
+    //val terminal = TerminalBuilder.builder().build()
+    val terminal  = TerminalBuilder.terminal() // defaults for now
+    val completer = new Completion(completion)
+    val parser    = new ReplParser(repl)
+    val history   = new DefaultHistory
+
+    val builder =
+      LineReaderBuilder.builder()
+      .appName("scala")
+      .completer(completer)
+      .history(history)
+      .parser(parser)
+      .terminal(terminal)
+
+    locally {
+      import LineReader._, Option._
+      builder
+        .option(AUTO_GROUP, false)
+        .option(LIST_PACKED, true)  // TODO
+        .option(INSERT_TAB, true)   // At the beginning of the line, insert tab instead of completing
+        .option(HISTORY_TIMESTAMPED, false)         // Allows sharing history file between versions
+        .variable(HISTORY_FILE, config.historyFile) // Save history to file
+        .variable(SECONDARY_PROMPT_PATTERN, config.encolor(config.continueText)) // Continue prompt
+    }
+
+    val reader = builder.build()
+    locally {
+      import LineReader._, Option._
+      // VIINS, VICMD, EMACS
+      val keymap = if (config.viMode) VIINS else EMACS
+      reader.getKeyMaps.put(MAIN, reader.getKeyMaps.get(keymap));
+    }
+    def backupHistory(): Unit = {
+      import java.nio.file.{Files, Paths, StandardCopyOption}, StandardCopyOption.REPLACE_EXISTING
+      val hf = Paths.get(config.historyFile)
+      val bk = Paths.get(config.historyFile + ".bk")
```

Comment:
When creating the history file, we [change its permissions](https://github.com/scala/scala/blob/515916c42f8f2f5c476ffabd6774e5a903feda10/src/repl-jline/scala/tools/nsc/interpreter/jline/FileBackedHistory.scala#L32-L44) to `600`. This is a security measure for multi-user systems that don't secure the `$HOME` by default (I'm looking at you OS/X and Windows).

 We should do the same with the backup and the new copy of the history file.

Reply:
spun out as https://github.com/scala/scala-dev/issues/684